tracer/src/Datadog.Trace/Tagging/TagPropagation.cs

```diff
@@ -0,0 +1,208 @@
+                        cachedHeader = null;
+                    }
+                }
+                else
+                {
+                    // we can't reuse the same string if we skip any key/value pair
+                    cachedHeader = null;
+                }
+            }
+            else
+            {
+                // we can't reuse the same string if we skip any key/value pair
+                cachedHeader = null;
+            }
+        }
+
+        return traceTags.Count > 0 ? new TraceTagCollection(traceTags, cachedHeader) : new TraceTagCollection();
+    }
+
+    /// <summary>
+    /// Constructs a string that can be used for horizontal propagation using the "x-datadog-tags" header
+    /// in a "key1=value1,key2=value2" format. This header should only include tags with the "_dd.p.*" prefix.
+    /// The returned string is cached and reused if no relevant tags are changed between calls.
+    /// </summary>
+    /// <returns>A string that can be used for horizontal propagation using the "x-datadog-tags" header.</returns>
+    public static string ToHeader(TraceTagCollection tags, int maxOutgoingHeaderLength)
+    {
+        if (tags.Count == 0)
+        {
+            return string.Empty;
+        }
+
+        var sb = StringBuilderCache.Acquire(StringBuilderCache.MaxBuilderSize);
+
+        foreach (var tag in tags)
+        {
+            if (!string.IsNullOrEmpty(tag.Key) &&
+                !string.IsNullOrEmpty(tag.Value) &&
+                tag.Key.StartsWith(PropagatedTagPrefix, StringComparison.Ordinal))
+            {
+                if (!IsValid(tag.Key, tag.Value))
+                {
+                    Log.Debug("Propagated tag is not valid. Key: \"{key}\", Value: \"{value}\"", tag.Key, tag.Value);
+
+                    // if tag contains invalid chars,
+                    // set tag "_dd.propagation_error:encoding_error"...
+                    tags.SetTag(Tags.TagPropagation.Error, PropagationErrorTagValues.EncodingError);
+
+                    // ... and don't set the header
+                    return string.Empty;
+                }
+
+                if (sb.Length > 0)
+                {
+                    sb.Append(TagPairSeparator);
+                }
+
+                sb.Append(tag.Key)
+                  .Append(KeyValueSeparator)
+                  .Append(tag.Value);
+            }
+
+            if (sb.Length > maxOutgoingHeaderLength)
+            {
+                Log.Debug<int, int>("Outgoing tag propagation header is too long. Length: {0}, Maximum: {1}.", sb.Length, maxOutgoingHeaderLength);
+
+                // if combined tags get too long for propagation headers,
+                // set tag "_dd.propagation_error:inject_max_size"...
+                tags.SetTag(Tags.TagPropagation.Error, PropagationErrorTagValues.InjectMaxSize);
+
+                // ... and don't set the header
+                return string.Empty;
```

Comment:
Oh good spot Pierre! 🔍 Another idea that I had is if we should create a lightweight struct that implements `IDisposable`, that way we could use the `using` statement with it. Maybe that would be doing too much, just thought I'd throw it out there

Reply:
I created a small PR to implement the release method since I noticed a couple other places with this same behavior. Once it's merged and you rebase on top of it, you can call `StringBuilderCache.Release(StringBuilder)` here to release the cached instance.

See #2956